<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Citations and references &#8212; cimr 0.1.5 documentation</title>
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Feature-based analyses using variant-association files" href="feature_analyses/gene.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          cimr</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.5</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="https://gihub.com/greenelab/cimr.html">github</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><p class="caption"><span class="caption-text">cimr guides</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="install.html#installing-python-and-git-lfs">Installing Python and git(-lfs)</a></li>
<li class="toctree-l2"><a class="reference internal" href="install.html#installing-cimr">Installing cimr</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How to contribute to cimr</a><ul>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#did-you-find-a-bug">Did you find a bug?</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#did-you-fix-a-bug">Did you fix a bug?</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#do-you-have-data">Do you have data?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="processing/variant.html">Processing variant-association files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#input-file-examples">Input file examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#output-file-examples">Output (file) examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#parameters-and-defaults">Parameters and defaults</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#example-usage">Example usage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="processing/data.html">Pre-processed data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="processing/data.html#genome-wide-association-studies">Genome-wide association studies</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="feature_analyses/gene.html">Feature-based analyses using variant-association files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="feature_analyses/gene.html#gene-based-association-tests">Gene-based association tests</a></li>
<li class="toctree-l2"><a class="reference internal" href="feature_analyses/gene.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Citations and references</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#cimr-and-cimr-d">cimr and cimr-d</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genome-wide-association-studies">Genome-wide association studies</a></li>
<li class="toctree-l2"><a class="reference internal" href="#expression-quantitative-trait-loci-eqtl-mapping">expression quantitative trait loci (eQTL) mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotations">Annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#d-genome-organization">3D genome organization</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Citations and references</a><ul>
<li><a class="reference internal" href="#cimr-and-cimr-d">cimr and cimr-d</a></li>
<li><a class="reference internal" href="#genome-wide-association-studies">Genome-wide association studies</a></li>
<li><a class="reference internal" href="#expression-quantitative-trait-loci-eqtl-mapping">expression quantitative trait loci (eQTL) mapping</a><ul>
<li><a class="reference internal" href="#genotype-tissue-expression-project-gtex">Genotype-Tissue Expression project (GTEx)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#annotations">Annotations</a></li>
<li><a class="reference internal" href="#d-genome-organization">3D genome organization</a><ul>
<li><a class="reference internal" href="#tad-coordinates">TAD coordinates</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="feature_analyses/gene.html" title="Previous Chapter: Feature-based analyses using variant-association files"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; Feature-based...</span>
    </a>
  </li>
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/citations.rst.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="section" id="citations-and-references">
<h1>Citations and references<a class="headerlink" href="#citations-and-references" title="Permalink to this headline">¶</a></h1>
<p>This document contains citation information for all data, method, or any
relevant information used to make cimr work.</p>
<div class="section" id="cimr-and-cimr-d">
<h2>cimr and cimr-d<a class="headerlink" href="#cimr-and-cimr-d" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>If you find cimr and cimr-d useful, please cite:</p>
<ul>
<li><p>Park Y. &amp; Greene C., (2019). Continuous integration and analysis using cimr,
In Preparation</p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="genome-wide-association-studies">
<h2>Genome-wide association studies<a class="headerlink" href="#genome-wide-association-studies" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>The GWAS findings used in cimr analysis are originally published in the
following articles:</p></li>
</ul>
</div>
<div class="section" id="expression-quantitative-trait-loci-eqtl-mapping">
<h2>expression quantitative trait loci (eQTL) mapping<a class="headerlink" href="#expression-quantitative-trait-loci-eqtl-mapping" title="Permalink to this headline">¶</a></h2>
<div class="section" id="genotype-tissue-expression-project-gtex">
<h3>Genotype-Tissue Expression project (GTEx)<a class="headerlink" href="#genotype-tissue-expression-project-gtex" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The article containing the latest GTEx main analysis results:</p>
<ul>
<li><p><a class="reference external" href="https://www.nature.com/articles/nature24277">The GTEx Consortium (2017)</a>
Genetic effects on gene expression across human tissues. Nature. 550: 204–213</p></li>
<li><p>The full summary-level data are on <a class="reference external" href="https://gtexportal.org">GTEx Portal</a></p></li>
<li><p>The raw and individualized data are on
dbGaP <a class="reference external" href="https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000424.v7.p2">phs000424.v7.p2</a></p></li>
</ul>
</li>
<li><p>The article containing the eQTLGen main analysis results:</p></li>
</ul>
</div>
</div>
<div class="section" id="annotations">
<h2>Annotations<a class="headerlink" href="#annotations" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="d-genome-organization">
<h2>3D genome organization<a class="headerlink" href="#d-genome-organization" title="Permalink to this headline">¶</a></h2>
<div class="section" id="tad-coordinates">
<h3>TAD coordinates<a class="headerlink" href="#tad-coordinates" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The article containing the method used to call TAD boundaries:</p>
<ul>
<li><p>Wang Y., Song F., Zhang B., Zhang L., Xu J., Kuang D., Li D., Choudhary M.N.K.,
Li Y., Hu M.,  Hardison R., Wang T. &amp; Yue F. (2018).
The 3D Genome Browser: a web-based browser for visualizing 3D genome
organization and long-range chromatin interactions. Genome Biology. 19, 151.
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/30286773">pubmed</a> |
<a class="reference external" href="http://promoter.bx.psu.edu/hi-c/downloads/">dataset</a></p></li>
</ul>
</li>
<li><p>The TAD coordinates are originally published in the following articles:</p>
<ul>
<li><p>Dixon J.R., Selvaraj S., Yue F., Kim A., Li Y., Shen Y., Hu M., Liu J.S. &amp; Ren B.
Topological domains in mammalian genomes identified by analysis of chromatin
interactions. Nature. 2012;485(7398):376-380.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/22495300">pubmed</a> |
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE35156">dataset</a></p></li>
<li><p>Dixon J.R., Jung I., Selvaraj S., Shen Y., Antosiewicz-Bourget J.E., Lee A.Y.,
Ye Z., Kim A., Rajagopal N., Xie W., Diao Y., Liang J., Zhao H., Lobanenkov V.V.,
Ecker J.R., Thomson J.A. &amp; Ren B. Chromatin architecture reorganization during
stem cell differentiation. Nature. 2015;518(7539):331-336
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/25693564">pubmed</a> |
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52457">dataset</a></p></li>
<li><p>ENCODE Project Consortium. An integrated encyclopedia of DNA elements in the
human genome. Nature. 2012 Sep 6;489(7414):57-74.
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/22955616">pubmed</a> |
<a class="reference external" href="https://www.encodeproject.org/search/?type=Experiment&amp;assay_title=Hi-C&amp;status=released&amp;limit=all">dataset</a></p></li>
<li><p>Leung D., Jung I., Rajagopal N., Schmitt A., Selvaraj S., Lee A. Y. &amp; Ren B.
Integrative analysis of haplotype-resolved epigenomes across human tissues.
Nature. 2015;518(7539):350-354
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/25693566">pubmed</a> |
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE58752">dataset</a></p></li>
<li><p>Lieberman-Aiden E., Nynke L. van Berkum, Williams L., Imakaev M., Ragoczy T.,
Telling A., Amit I., Lajoie B.R., Sabo P.J., Dorschner M.O., Sandstrom R.,
Bernstein B., Bender M.A., Groudine M., Gnirke A., Stamatoyannopoulos J.,
Mirny L.A., Lander E.S. &amp; Dekker J. Comprehensive Mapping of Long-Range
Interactions Reveals Folding Principles of the Human Genome.
Science. 2009. 326(5950):289-293.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/19815776">pubmed</a> |
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE18199">dataset</a></p></li>
<li><p>Rubin A., Barajas B., Furlan-Magaril M., Lopez-Pajares V., Mumbach M., Howard I.,
Kim D.S., Boxer L.D., Cairns J., Spivakov M., Wingett S.W., Shi M., Zhao Z.,
Greenleaf W.J., Kundaje A., Snyder M., Chang H., Fraser P., &amp; Khavari P.
Lineage-specific dynamic and pre-established enhancer-promoter contacts
cooperate in terminal differentiation. Nat. Genet 2017 , 49(10), 1522-1528.
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/28805829">pubmed</a> |
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84660">dataset</a></p></li>
<li><p>Schmitt A., Hu M., Jung I., Xu Z., Qiu Y., Tan C., Li Y., Lin S., Lin Y.,
Barr C. &amp; Ren B. A Compendium of Chromatin Contact Maps Reveals Spatially
Active Regions in the Human Genome. Cell Reports. 2016;17(8):2042-2059.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/27851967">pubmed</a> |
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87112">dataset</a></p></li>
</ul>
</li>
</ul>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2018-2019, YoSon Park.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.1.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>