<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>How to contribute to cimr &#8212; cimr 0.1.5 documentation</title>
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Processing variant-association files" href="processing/variant.html" />
    <link rel="prev" title="Installation" href="install.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          cimr</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.5</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="https://gihub.com/greenelab/cimr.html">github</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><p class="caption"><span class="caption-text">cimr guides</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="install.html#installing-python-and-git-lfs">Installing Python and git(-lfs)</a></li>
<li class="toctree-l2"><a class="reference internal" href="install.html#installing-cimr">Installing cimr</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">How to contribute to cimr</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#did-you-find-a-bug">Did you find a bug?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#did-you-fix-a-bug">Did you fix a bug?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#do-you-have-data">Do you have data?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="processing/variant.html">Processing variant-association files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#input-file-examples">Input file examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#output-file-examples">Output (file) examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#parameters-and-defaults">Parameters and defaults</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#example-usage">Example usage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="processing/data.html">Pre-processed data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="processing/data.html#genome-wide-association-studies">Genome-wide association studies</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="feature_analyses/gene.html">Feature-based analyses using variant-association files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="feature_analyses/gene.html#gene-based-association-tests">Gene-based association tests</a></li>
<li class="toctree-l2"><a class="reference internal" href="feature_analyses/gene.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations and references</a><ul>
<li class="toctree-l2"><a class="reference internal" href="citations.html#cimr-and-cimr-d">cimr and cimr-d</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#genome-wide-association-studies">Genome-wide association studies</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#expression-quantitative-trait-loci-eqtl-mapping">expression quantitative trait loci (eQTL) mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#annotations">Annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#d-genome-organization">3D genome organization</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">How to contribute to cimr</a><ul>
<li><a class="reference internal" href="#did-you-find-a-bug">Did you find a bug?</a></li>
<li><a class="reference internal" href="#did-you-fix-a-bug">Did you fix a bug?</a></li>
<li><a class="reference internal" href="#do-you-have-data">Do you have data?</a><ul>
<li><a class="reference internal" href="#variant-based-association-study-results">Variant-based association study results</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="install.html" title="Previous Chapter: Installation"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; Installation</span>
    </a>
  </li>
  <li>
    <a href="processing/variant.html" title="Next Chapter: Processing variant-association files"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">Processing va... &raquo;</span>
    </a>
  </li>
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/contributing.md.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="section" id="how-to-contribute-to-cimr">
<h1>How to contribute to cimr<a class="headerlink" href="#how-to-contribute-to-cimr" title="Permalink to this headline">¶</a></h1>
<p>cimr is an open source project that welcomes contributions from the
community, including:</p>
<ul class="simple">
<li><p>Data contributions</p>
<ul>
<li><p>Variant association summary statistics</p></li>
<li><p>Annotations and references</p></li>
</ul>
</li>
<li><p>Method contributions</p></li>
<li><p>Code contributions</p>
<ul>
<li><p>Code patches,</p></li>
<li><p>Documentation improvements,</p></li>
<li><p><a class="reference external" href="https://github.com/greenelab/cimr/issues">Bug reports</a></p></li>
<li><p>Reviews for contributed patches</p></li>
</ul>
</li>
</ul>
<div class="section" id="did-you-find-a-bug">
<h2>Did you find a bug?<a class="headerlink" href="#did-you-find-a-bug" title="Permalink to this headline">¶</a></h2>
<p>Ensure the bug was not already reported by searching on
<a class="reference external" href="https://github.com/greenelab/cimr/issues">github issues</a>.
If there are no open issues addressing the problem(s), open a
new one. Be sure to clearly state a title, description and as much
relevant information as possible with example codes and logs.</p>
</div>
<div class="section" id="did-you-fix-a-bug">
<h2>Did you fix a bug?<a class="headerlink" href="#did-you-fix-a-bug" title="Permalink to this headline">¶</a></h2>
<p>Open a new github pull request (PR) with the patch. Ensure that the
PR describes the problem and proposed solution. Refer to relevant
issues with a link. PR regarding cosmetic or stylistic edits are not
accepted.</p>
</div>
<div class="section" id="do-you-have-data">
<h2>Do you have data?<a class="headerlink" href="#do-you-have-data" title="Permalink to this headline">¶</a></h2>
<p>Ensure that the data is publicly available or contains a license to
share with attribution (e.g. CC BY 4.0). Briefly, any variant- or
gene-based association summary statistics passing the automated checks
via cimr may be elligible to be integrated into cimr-d, which would
allow faster, more convenient use by others in the community.</p>
<p>All due credits, acknowledgements or citations are recommended for
users of this resource. cimr-d is provided as a public repository via
git. Information regarding versioned archives for data freezes will be
provided as they become available. License or other permissions may be
necessary to integrate into cimr-d.</p>
<div class="section" id="variant-based-association-study-results">
<h3>Variant-based association study results<a class="headerlink" href="#variant-based-association-study-results" title="Permalink to this headline">¶</a></h3>
<p>Genome-wide association study (GWAS) and expression- or other
quantitative trait loci (eQTL, pQTL, sQTL, TWAS, etc.) have similar
input formatting requirements.</p>
<p>Following are required for all association summary statistics:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rsnum</span> <span class="p">:</span> <span class="n">rs</span> <span class="nb">id</span> <span class="n">of</span> <span class="n">the</span> <span class="n">variant</span>
<span class="n">variant_id</span> <span class="p">:</span> <span class="n">variant</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">following</span> <span class="nb">format</span>
    <span class="n">chromosome_position_referenceallele_alternateallele_genomebuild</span>
<span class="n">pval</span> <span class="o">/</span> <span class="n">pvalue</span><span class="p">:</span> <span class="n">p</span><span class="o">-</span><span class="n">value</span> <span class="n">of</span> <span class="n">the</span> <span class="n">beta</span> <span class="n">estimate</span>
<span class="n">beta</span> <span class="o">/</span> <span class="n">effect_size</span> <span class="p">:</span> <span class="n">beta</span> <span class="n">coefficient</span> <span class="n">estimate</span> <span class="k">for</span> <span class="n">the</span> <span class="n">effect</span>
    <span class="n">of</span> <span class="n">the</span> <span class="n">variant</span> <span class="n">on</span> <span class="n">the</span> <span class="n">gene</span>
<span class="n">se</span> <span class="o">/</span> <span class="n">standard_error</span> <span class="p">:</span> <span class="n">standard</span> <span class="n">error</span> <span class="n">of</span> <span class="n">the</span> <span class="n">beta</span>
<span class="n">zscore</span><span class="p">:</span> <span class="n">standardized</span> <span class="ow">or</span> <span class="n">imputed</span> <span class="n">statistic</span>
<span class="n">sample_size</span> <span class="p">:</span> <span class="n">sample</span> <span class="n">size</span> <span class="k">for</span> <span class="n">the</span> <span class="n">variant</span> <span class="n">underlying</span> <span class="n">the</span>
    <span class="n">effect_size</span> <span class="ow">and</span> <span class="n">pvalue</span> <span class="n">estimates</span><span class="o">.</span> <span class="n">If</span> <span class="ow">not</span> <span class="n">available</span> <span class="n">per</span><span class="o">-</span><span class="n">variant</span><span class="p">,</span>
    <span class="n">the</span> <span class="n">total</span> <span class="n">sample</span> <span class="n">size</span> <span class="n">used</span> <span class="k">for</span> <span class="n">the</span> <span class="n">study</span> <span class="ow">is</span> <span class="n">accepted</span><span class="o">.</span>
<span class="n">n_cases</span> <span class="p">:</span> <span class="o">**</span><span class="n">required</span> <span class="k">for</span> <span class="n">binary</span> <span class="n">phenotypic</span> <span class="n">traits</span> <span class="n">only</span><span class="o">.**</span>
    <span class="n">n_cases</span> <span class="n">are</span> <span class="n">necessary</span>
    <span class="k">for</span> <span class="n">downstream</span> <span class="n">analyses</span> <span class="n">such</span> <span class="k">as</span> <span class="n">coloc</span> <span class="n">colocalization</span><span class="o">.</span>
<span class="n">build</span><span class="p">:</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">included</span> <span class="ow">in</span> <span class="n">variant_id</span><span class="p">,</span> <span class="n">genome</span> <span class="n">build</span> <span class="n">can</span> <span class="n">be</span> <span class="n">provided</span>
    <span class="k">as</span> <span class="n">a</span> <span class="n">yaml</span> <span class="n">parameter</span>
</pre></div>
</div>
<p>Following are required for quantitative trait loci:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">feature_id</span> <span class="p">:</span> <span class="n">feature</span> <span class="n">may</span> <span class="n">be</span> <span class="n">gene</span> <span class="p">(</span><span class="n">eQTL</span><span class="p">),</span> <span class="n">transcript</span> <span class="p">(</span><span class="n">sQTL</span><span class="p">)</span>
</pre></div>
</div>
<p>Following are optional:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tss_distance</span> <span class="p">:</span> <span class="n">distance</span> <span class="n">to</span> <span class="n">the</span> <span class="n">transcription</span> <span class="n">start</span> <span class="n">site</span> <span class="n">of</span> <span class="n">the</span>
    <span class="n">gene_id</span><span class="p">,</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="n">used</span> <span class="k">as</span> <span class="n">a</span> <span class="n">feature</span> <span class="p">(</span><span class="n">eQTL</span><span class="p">)</span>
<span class="n">ma_samples</span> <span class="p">:</span> <span class="n">samples</span> <span class="k">with</span> <span class="n">minor</span> <span class="n">alleles</span>
<span class="n">maf</span> <span class="p">:</span> <span class="n">minor</span> <span class="n">allele</span> <span class="n">frequency</span>
<span class="n">effect_allele</span> <span class="p">:</span> <span class="n">allele</span> <span class="k">with</span> <span class="n">respect</span> <span class="n">to</span> <span class="n">which</span> <span class="n">variant</span><span class="s1">&#39;s effect</span>
    <span class="n">sizes</span> <span class="n">are</span> <span class="n">estimated</span>
<span class="n">non_effect_allele</span> <span class="p">:</span> <span class="n">non</span><span class="o">-</span><span class="n">effect</span><span class="o">-</span><span class="n">allele</span>
<span class="n">imputation_status</span> <span class="p">:</span> <span class="n">imputation</span> <span class="n">status</span> <span class="n">of</span> <span class="n">the</span> <span class="n">variant</span><span class="o">.</span> <span class="s1">&#39;original&#39;</span>
    <span class="n">to</span> <span class="n">indicate</span> <span class="n">that</span> <span class="n">the</span> <span class="n">original</span> <span class="n">GWAS</span> <span class="n">contained</span> <span class="n">the</span> <span class="n">variant</span><span class="o">.</span>
    <span class="s1">&#39;imputed&#39;</span> <span class="n">to</span> <span class="n">indicate</span> <span class="n">that</span> <span class="n">the</span> <span class="n">data</span> <span class="n">submitter</span> <span class="n">has</span> <span class="n">done</span>
    <span class="p">(</span><span class="n">additional</span><span class="p">)</span> <span class="n">imputation</span> <span class="n">to</span> <span class="n">the</span> <span class="n">initial</span> <span class="n">release</span> <span class="n">of</span> <span class="n">the</span> <span class="n">data</span><span class="o">.</span>
<span class="n">frequency</span> <span class="p">:</span> <span class="n">allele</span> <span class="n">frequency</span> <span class="n">of</span> <span class="n">the</span> <span class="n">effect_allele</span>
</pre></div>
</div>
<p>Following are assumed by parsing variant_id column:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chrom</span> <span class="p">:</span> <span class="n">chromosome</span> <span class="nb">id</span>
<span class="n">pos</span><span class="p">:</span> <span class="n">genomic</span> <span class="n">position</span>
<span class="n">ref</span> <span class="p">:</span> <span class="n">reference</span> <span class="n">allele</span>
<span class="n">alt</span> <span class="p">:</span> <span class="n">alternate</span> <span class="n">allele</span>
<span class="n">build</span> <span class="p">:</span> <span class="n">genomic</span> <span class="n">build</span> <span class="n">version</span> <span class="n">number</span> <span class="k">for</span> <span class="n">the</span> <span class="n">variant_id</span>
</pre></div>
</div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2018-2019, YoSon Park.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.1.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>