<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>cimr: data analysis toolkit using association mapping &#8212; cimr 0.1.5 documentation</title>
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="install.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="#">
          cimr</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.5</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="https://gihub.com/greenelab/cimr.html">github</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><p class="caption"><span class="caption-text">cimr guides</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="install.html#installing-python-and-git-lfs">Installing Python and git(-lfs)</a></li>
<li class="toctree-l2"><a class="reference internal" href="install.html#installing-cimr">Installing cimr</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How to contribute to cimr</a><ul>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#did-you-find-a-bug">Did you find a bug?</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#did-you-fix-a-bug">Did you fix a bug?</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#do-you-have-data">Do you have data?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="processing/variant.html">Processing variant-association files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#input-file-examples">Input file examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#output-file-examples">Output (file) examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#parameters-and-defaults">Parameters and defaults</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#example-usage">Example usage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="processing/data.html">Pre-processed data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="processing/data.html#genome-wide-association-studies">Genome-wide association studies</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="feature_analyses/gene.html">Feature-based analyses using variant-association files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="feature_analyses/gene.html#gene-based-association-tests">Gene-based association tests</a></li>
<li class="toctree-l2"><a class="reference internal" href="feature_analyses/gene.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations and references</a><ul>
<li class="toctree-l2"><a class="reference internal" href="citations.html#cimr-and-cimr-d">cimr and cimr-d</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#genome-wide-association-studies">Genome-wide association studies</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#expression-quantitative-trait-loci-eqtl-mapping">expression quantitative trait loci (eQTL) mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#annotations">Annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#d-genome-organization">3D genome organization</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">cimr: data analysis toolkit using association mapping</a></li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="install.html" title="Next Chapter: Installation"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">Installation &raquo;</span>
    </a>
  </li>
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/index.rst.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <span class="target" id="module-cimr"></span><div class="section" id="cimr-data-analysis-toolkit-using-association-mapping">
<h1>cimr: data analysis toolkit using association mapping<a class="headerlink" href="#cimr-data-analysis-toolkit-using-association-mapping" title="Permalink to this headline">Â¶</a></h1>
<p><strong>Date:</strong> Aug 21, 2019 <strong>Version:</strong> 0.1.5</p>
<p><a class="reference external" href="https://github.com/greenelab/cimr">cimr</a> is an open source, BSD-licensed
library providing data analysis tool for the <a class="reference external" href="https://www.python.org/">Python</a>
programming language. cimr primarily utilizes publicly available summary
statistics files from genome-wide association study (GWAS), expression quantitative
trait loci (eQTL) or other variant-based association mapping results.
For contributing new data into cimr-d, see
<a class="reference external" href="https://github.com/greenelab/cimr-d/blob/master/doc/contributing.md">contributions document</a> .
cimr also offers some utility functions to process variant-based association
summary statistics and to conduct gene-based tests.</p>
<p><strong>Useful links</strong>:
<a class="reference external" href="https://github.com/greenelab/cimr">Source repository</a> |
<a class="reference external" href="https://github.com/greenelab/cimr/issues">Issues &amp; Ideas</a></p>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">cimr guides</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="install.html#installing-python-and-git-lfs">Installing Python and git(-lfs)</a></li>
<li class="toctree-l2"><a class="reference internal" href="install.html#installing-cimr">Installing cimr</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How to contribute to cimr</a><ul>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#did-you-find-a-bug">Did you find a bug?</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#did-you-fix-a-bug">Did you fix a bug?</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#do-you-have-data">Do you have data?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="processing/variant.html">Processing variant-association files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#input-file-examples">Input file examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#output-file-examples">Output (file) examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#parameters-and-defaults">Parameters and defaults</a></li>
<li class="toctree-l2"><a class="reference internal" href="processing/variant.html#example-usage">Example usage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="processing/data.html">Pre-processed data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="processing/data.html#genome-wide-association-studies">Genome-wide association studies</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="feature_analyses/gene.html">Feature-based analyses using variant-association files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="feature_analyses/gene.html#gene-based-association-tests">Gene-based association tests</a></li>
<li class="toctree-l2"><a class="reference internal" href="feature_analyses/gene.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations and references</a><ul>
<li class="toctree-l2"><a class="reference internal" href="citations.html#cimr-and-cimr-d">cimr and cimr-d</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#genome-wide-association-studies">Genome-wide association studies</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#expression-quantitative-trait-loci-eqtl-mapping">expression quantitative trait loci (eQTL) mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#annotations">Annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="citations.html#d-genome-organization">3D genome organization</a></li>
</ul>
</li>
</ul>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2018-2019, YoSon Park.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.1.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>